### PR TITLE
Task/3008 get registrations

### PR DESCRIPTION
**Pull request by kzangeli**
Implemented issue #3008.
GET Registration in APIv2
Not 100% complete functional test as not everything can be tested due to the lack of registration creation in APIv2.
More functests to be added once we implement `POST /v2/registration`.

### CODE REVIEW

File: src/lib/apiTypesV2/Registration.cpp [hunk 1]
```diff
@@ -71,10 +119,15 @@ std::string DataProvided::toJson(void)
 *
 * Provider::toJson -
 */
-std::string Provider::toJson()
+std::string Provider::toJson(void)
 {
-  JsonHelper jh;
-  // TBD
+  JsonHelper   jh;
+  std::string  urlAsJson = "{\"url\": \"" + url + "\"}";
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Registration.cpp]: This is a bit ugly but ... url is a simple std::string.
I am not very familiar with JsonHelper but I believe we would need to create a new class for `url` to avoid this problem. Not sure it's worth it ...

File: src/lib/apiTypesV2/Registration.h [hunk 2]
```diff
@@ -25,74 +25,79 @@
...
+    "status": "failed",
+    "forwardingInformation": {
+      "timesSent": 12,
+      "lastForwarding": "2017-10-06T16:00:00.00Z",
+      "lastFailure": "2017-10-06T16:00:00.00Z",
+      "lastSuccess": "2017-10-05T18:25:00.00Z",
+    }
+  }
+------------------------------------------------------------------------------------
+#endif
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Registration.h]: Decided to keep this comment until we have registry creation in APIv2

File: src/lib/apiTypesV2/Registration.cpp (hunk 1 above)
Comment (fgalan) [reply src/lib/apiTypesV2/Registration.cpp]: I think the "canonical" solution would be to have a second JsonHelper and do something likes this:

```
JsonHelper jh2;
jh2.addString("url", url);

jh.addRaw("http", jh2.str());
```

The more "styled" solution would be that but, in this particular case, using `urlAsJson` string as it is now is also fine from my point of view. You chose, either "Fixed in" or "NTC" :)

File: src/lib/apiTypesV2/Registration.h [hunk 3]
```diff
@@ -25,74 +25,79 @@
...
+
 /* ****************************************************************************
 *
 * ForwardingInformation -
 */
-struct ForwardingInformation
+class ForwardingInformation
 {
-  int          lastFailure;  // FIXME: in Subscrition.h we use int for this, not long long as in lastNotification. why?
-  int          lastSuccess;  // FIXME: in Subscrition.h we use int for this, not long long as in lastNotification. why?
```
Comment (fgalan) [inline src/lib/apiTypesV2/Registration.h]: No clue... maybe we didn't realize at the very begining and we haven't change for the fear of breaking backward compability at DB level? The change shouldn't hurt but I'm not fully sure...

File: src/lib/apiTypesV2/Subscription.h [hunk 4]
```diff
@@ -75,8 +75,8 @@ struct Notification
   long long                lastNotification;
   HttpInfo                 httpInfo;
   std::string              toJson(const std::string& attrsFormat);
-  int                      lastFailure;
-  int                      lastSuccess;
+  int                      lastFailure;  // FIXME P4: should be long long
```
Comment (fgalan) [inline src/lib/apiTypesV2/Subscription.h]: I'd suggest add the ending "... as lastNotification is"

File: src/lib/common/JsonHelper.cpp [hunk 5]
```diff
@@ -215,19 +220,38 @@ void JsonHelper::addNumber(const std::string& key, long long value)
   empty = false;
 }
 
+
+
+/* ****************************************************************************
+*
+* JsonHelper::addInt -
+*/
+void JsonHelper::addInt(const std::string& key, long long value)
```
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: Just to confirm... the only actual changes in this files are addInt() and addBoolean() functions, all the rest is style and typo fixing. Is correct?

File: src/lib/common/JsonHelper.cpp [hunk 6]
```diff
@@ -215,19 +220,38 @@ void JsonHelper::addNumber(const std::string& key, long long value)
...
+{
+  if (!empty)
```
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: This function is exactly equal to addNumber(). Why don't use the existing one?

File: src/lib/common/JsonHelper.cpp [hunk 7]
```diff
@@ -254,6 +280,17 @@ void JsonHelper::addDate(const std::string& key, long long timestamp)
 
 
 
+/* ****************************************************************************
+*
+* JsonHelper::addBool -
+*/
+void JsonHelper::addBool(const std::string& key, bool b)
```
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: First time we need to add booleans to rendered JSON output in history? :)

The addition is great... just a bit surprised we didn't have it already.

File: src/lib/common/statistics.cpp [hunk 8]
```diff
@@ -159,6 +159,7 @@ int noOfIndividualSubscriptionRequest                    = -1;
 int noOfSimulatedNotifications                           = -1;
 int noOfBatchQueryRequest                                = -1;
 int noOfBatchUpdateRequest                               = -1;
+int noOfRegistrationRequest                              = -1;
```
Comment (fgalan) [inline src/lib/common/statistics.cpp]: This involves a new item in the GET /stats (?) operation? Have it any impact in documentation?

File: src/lib/mongoBackend/CMakeLists.txt [hunk 9]
```diff
@@ -43,7 +43,7 @@ SET (SOURCES
     TriggeredSubscription.cpp
     mongoConnectionPool.cpp
     mongoGetSubscriptions.cpp
-    mongoGetRegistrations.cpp
+    mongoRegistrationGet.cpp
```
Comment (fgalan) [inline src/lib/mongoBackend/CMakeLists.txt]: For "direct" operations (I mean, the ones not based in a mapping to existing NGSIv1 based operation) I think the naming criteria is "mongo" + verb + resource. Is this a causal change due to a matter of taste? Or it is strongly intentional?

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+
+
+
+/* ****************************************************************************
+*
+* setDataProvided -
+*
+* Make sure there is only ONE "contextRegistration" in the vector
+* If we have more than one, then the Registration is made in API V1 as this is not
+* possible in V2 and we cannot respond to the request using the current implementation of V2
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: I'd suggest to add a reference to issue https://github.com/telefonicaid/fiware-orion/issues/3044, i.e.

> This function will be change to work in different way upon completion of issue #3044

or something similar

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+  // Extract first CR from the contextRegistration vector
+  //
+  mongo::BSONObj cr0 = crV[0].embeddedObject();
+
+
+
+  //
+  // Entities
+  //
+  std::vector<mongo::BSONElement>  dbEntityV = getFieldF(cr0, REG_ENTITIES).Array();
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: Maybe the logic for "filling" entities could be done in a separated function, to make setDataProvider() shorter and easier.

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    regP->dataProvided.entities.push_back(entity);
+  }
+
+
+  //
+  // Attributes
+  //
+  std::vector<mongo::BSONElement> dbAttributeV = getFieldF(cr0, REG_ATTRS).Array();
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: Maybe the logic for "filling" attributes could be done in a separated function, to make setDataProvider() shorter and easier.

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+
+
+
+/* ****************************************************************************
+*
+* setForwardingInformation -
+*/
+static void setForwardingInformation(ngsiv2::Registration* regP, const mongo::BSONObj& r)
+{
+}
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: I'd suggest to add some brief explanation about why this function is empty, even a reference to issue if we have the explanation already in some existing issue.

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+  }
+
+  reqSemTake(__FUNCTION__, "Mongo Get Registration", SemReadOp, &reqSemTaken);
+
+  LM_T(LmtMongo, ("Mongo Get Registration"));
+
+  std::auto_ptr<mongo::DBClientCursor>  cursor;
+  mongo::BSONObj                        q;
+
+  if (servicePath.empty())
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: Looking to how this implemented in subscriptions:

```
BSONObj q     = BSON("_id" << oid);
```

(In other words, SP is not taken into account in the query)

So my nice theory about how SP works in subs (and should work in regs) exposed in https://github.com/telefonicaid/fiware-orion/issues/3051 colapses down :D

Thus, I propose the following:

1. To create an issue about "how SP should be interpreted in retrieve operations for subscriptions and registrations" (expaning #3051 or creating a new one)
2. Implement this by the moment with `q = BSON("_id" << oid);` as subscription are, pending to have a final decision of what to do in the issue.

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+*
+* mongoRegistrationGet - 
+*/
+void mongoRegistrationGet
+(
+  ngsiv2::Registration*  regP,
+  const std::string&     regId,
+  const std::string&     tenant,
+  const std::string&     servicePath,
+  OrionError*            oeP
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: Is our "style" to have OrionError always at the end in mongoBackend "external" functions?

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 14 above)
Comment (fgalan) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Well, not completely collapsed ;)... I think is stills making sense but for other operation.

It seems that SP is taken into account in "getting all" operation (i.e. GET /v2/subscrpitions, without id):

```
  if (!servicePath.empty() && servicePath != "/#")
  {
    q = Query(BSON(CSUB_SERVICE_PATH << servicePath));
  }
  else
  {
    q = Query();
  }
```

Maybe a good argument for not taking into account SP in GET /v2/subscriptions/id and GET /v2/registrations/id is that id univocally identifies the subscription/registration in that case, i.e. the "primary key" of the object doesn't include SP (as entities do).

File: src/lib/mongoBackend/mongoRegistrationGet.cpp [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+
+    //
+    // Fill in the Registration with data retrieved from the data base
+    //
+    setRegistrationId(regP, r);
+    setDescription(regP, r);
+    if (setDataProvided(regP, r, false) == false)
+    {
+      releaseMongoConnection(connection);
+      LM_W(("Bad Input (more than one registration, API V2 can't deal with that)"));
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoRegistrationGet.cpp]: Change Suggestion:

```
("Bad Input (getting registrations with more than one CR is not yet implemented, see issue 3044")
```

File: src/lib/serviceRoutinesV2/getRegistration.cpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+#include "serviceRoutinesV2/getRegistration.h"
+#include "alarmMgr/alarmMgr.h"
+
+
+
+/* ****************************************************************************
+*
+* getRegistration - 
+*
+* GET /v2/registration/<reg id>
```
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getRegistration.cpp]: GET /v2/registrations/<reg id>

(with 's' in the URL)

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+            }
+        ]
+    },
+    "forwardingInformation": {
+        "timesSent": 0
+    },
+    "id": "REGEX([0-9a-f\-]{24})",
+    "provider": {
+        "http": {
+            "url": "http://localhost:9801/v2"
```
Comment (fgalan) [inline test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: 9801 -> REGEX(\d+)

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+                "type": "T1"
+            },
+            {
+                "id": "E2",
+                "type": "T2"
+            }
+        ]
+    },
+    "forwardingInformation": {
+        "timesSent": 0
```
Comment (fgalan) [inline test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: Maybe you could add a test with a CPr and cause some forwarding in order to see how this timesSent increases. Makes sense?

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+--SHELL--
+
+#
+# At the moment of implementing the functionality of issue #3008, there was no way to create
+# registrations using APIv2 (Nov/Dec 2017), so, to test this new functionality of querying registrations
+# in APIv2, the registrations had to be created using APIv1.
+# Now, one might say, so, as soon as APIv2 implements creation of registrations, this test should be changed to
+# use APIv2 registration creation instead.
+# Sure, but, creating and getting registrations in pure APIv2 will be sufficiently tested in the future.
+# I would prefer to keep this tests as is, creation in APIv1 and getting in APIv2, to test the 'API version cross'.
```
Comment (fgalan) [inline test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: Yes, a test like this could be kept after POST /v2/registrations in order to check cross-version works.

However, I'd suggest to add a .test to check invalid id works fine, i.e. trying to GET /v2/registrations/<id> without previous creating anything and check the right error is returned.

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test (hunk 19 above)
Comment (fgalan) [reply test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: ...or maybe the forwarding logic has not been yet modified to increase timesSent each time a new query/update is forwarded?

File: src/lib/apiTypesV2/Registration.cpp (hunk 1 above)
Comment (kzangeli) [reply src/lib/apiTypesV2/Registration.cpp]: Ugly either way. Keeping the initial implementation.

NTC

File: src/lib/apiTypesV2/Registration.h (hunk 2 above)
Comment (kzangeli) [reply src/lib/apiTypesV2/Registration.h]: NTC

File: src/lib/common/JsonHelper.cpp (hunk 5 above)
Comment (kzangeli) [reply src/lib/common/JsonHelper.cpp]: Yes, that's it

File: src/lib/common/JsonHelper.cpp (hunk 6 above)
Comment (kzangeli) [reply src/lib/common/JsonHelper.cpp]: My idea is to remove addNumber in the end.
Number is either Float or Integer, and we have a method `addFloat` and not `addInteger` but instead `addNumber` ... Doesn't make any sense. Confusing,
So, I started this change here, adding the `addInt` method.
Creating an issue about removing `addNumber` in favor of `addInt`

File: src/lib/common/JsonHelper.cpp (hunk 7 above)
Comment (kzangeli) [reply src/lib/common/JsonHelper.cpp]: Yeah, strange, right?
Guess bools were added in some other way before. Haven't looked into it

File: src/lib/mongoBackend/CMakeLists.txt (hunk 9 above)
Comment (kzangeli) [reply src/lib/mongoBackend/CMakeLists.txt]: Actually, this is strongly intentional, as it follows the style guide (unfortunately it is just a Should ... https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/contribution_guidelines.md#s2-object-action-naming-convention).

The advantage with names like this is that all related files (mongoRegGet, mongoRegPost, mongoRegDelete, etc) are sorted together, instead of sorted by verb which is less interesting.
[ The file name is of course the same as the function name ].

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 11 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Good idea, let's do it !

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 14 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Yeah, it gets a bit tricky with the service path and the unique reg id ...
If you have the reg id ... should we even care about the service path?
I think I would vote against that, but open for discussion.

Let's decide something asap and I'll change this accordingly.

And yes, when more than one registration (without any reg id (GET /v2/registrations), the service path makes a lot of sense. Same for subs, of course.

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test (hunk 19 above)
Comment (kzangeli) [reply test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: I haven't lifted a finger yet about forwards,
This PR only deals with GETting registrations.
Once the V2 forwards are implemented, we'll implement this kind of tests

NTC (for this particualar PR)

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 12 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Definitely

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 15 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: We have no style about that, I'm afraid.
However, having this parameter last makes sense, at least to me.
First param is the receiver (output parameter), then the stuff to fill the output with, and the last param is part of the return value ...

I wouldn't want it to be mixed inside the input params, that would be strange.

Perhaps we should add something in the style guide ...

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 13 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Can't find an issue about API v2 forwarding.
There should be one, right?

File: src/lib/apiTypesV2/Subscription.h (hunk 4 above)
Comment (kzangeli) [reply src/lib/apiTypesV2/Subscription.h]: Fixed in f114c3e

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 10 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Fixed in f114c3e

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 11 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Fixed in f114c3e

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 12 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Fixed in f114c3e

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 13 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Fixed in f114c3e (but without issue)

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 16 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Fixed in f114c3e

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test (hunk 20 above)
Comment (kzangeli) [reply test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: Fixed in f114c3e

File: src/lib/common/statistics.cpp (hunk 8 above)
Comment (kzangeli) [reply src/lib/common/statistics.cpp]: Good comment, made me realize that the stats aren't complete.
When a new Service is defined in `enum RequestType`, the function `statisticsUpdate`, that has a switch statement with `enum RequestType` starts to complain.
To fix the warning, a new case is added with the counter for statistics of the new Service.

However, the counter has to be used as well, in a service routine called `renderCounterStats`.

Comment (kzangeli) [reply src/lib/common/statistics.cpp]: Fixed in e03804e (including the new counter in GET /statistics).

About documentation:
No, I think nothing is needed. We don't go into this kind of detail in the documentation.

File: src/lib/common/JsonHelper.cpp (hunk 5 above)
Comment (fgalan) [reply src/lib/common/JsonHelper.cpp]: NTC

File: src/lib/common/JsonHelper.cpp (hunk 6 above)
Comment (fgalan) [reply src/lib/common/JsonHelper.cpp]: So, after last comments in #3058 I understand this PR will be modified to use the existing version of addNumber(long long) and in a next and indepedent PR not related with NGSIv2-regs addFloat(float) will be changed to addNumber(float), isn't it?

File: src/lib/common/JsonHelper.cpp (hunk 7 above)
Comment (fgalan) [reply src/lib/common/JsonHelper.cpp]: NTC

File: src/lib/common/statistics.cpp (hunk 8 above)
Comment (fgalan) [reply src/lib/common/statistics.cpp]: I'm unsure about the impact in documentation... https://fiware-orion.readthedocs.io/en/master/admin/statistics/index.html section "Counter block" shows particular counters. However, this could be considered in the documentation as an example and not exahustive. Have a look to that piece of documentaiton and tell me what do you think, pls.

File: src/lib/mongoBackend/CMakeLists.txt (hunk 9 above)
Comment (fgalan) [reply src/lib/mongoBackend/CMakeLists.txt]: After reading S2 rule, it seems it refers to function and variable names... Does it also applied to file names? Maybe we should discuss this by skype.

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 13 above)
Comment (fgalan) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: > There should be one, right?

I think currently it isn't any issue about that. As much, some early dicussion in an old PR starting here: https://github.com/telefonicaid/fiware-orion/pull/2990#issuecomment-334726286. That could be reused to create the issue.

Any volunteer to create the issue? @jmcanterafonseca? ;)

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 15 above)
Comment (fgalan) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Maybe adding some rule about "first all the input parameters, next all the output parameters" ?

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 14 above)
Comment (fgalan) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: > Let's decide something asap and I'll change this accordingly, if necessary.

For this PR, ignore SP in the query and use just "q = BSON("_id" << oid);"

I'll take care of adjusting #3051 and add a piece of information about this.

File: src/lib/serviceRoutinesV2/getRegistration.cpp (hunk 17 above)
Comment (fgalan) [reply src/lib/serviceRoutinesV2/getRegistration.cpp]: Maybe you missed the "Fixed in" comment here?

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test (hunk 19 above)
Comment (fgalan) [reply test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: For this particular PR I'd suggest not rendering the `forwardingInformation` element. If we leave as it is now, we are adding wrong information, i.e. we always say `"timesSent": 0` even when some forwareded request has been done. Better say nothing than a lie, I guess.

File: src/lib/mongoBackend/CMakeLists.txt (hunk 9 above)
Comment (fgalan) [reply src/lib/mongoBackend/CMakeLists.txt]: Moreover, if at the end the file name it changed you should also do a similar change to this one in the SET(HEADERS, ...) part of the file.

Comment (fgalan) [reply src/lib/mongoBackend/CMakeLists.txt]: Morever bis :), the original mongoGetRegistrations.h file is still there. Mabye you copied it to create mongoRegistrationGet.h but didn't removed the original one. Please have a look.

File: src/lib/serviceRoutinesV2/getRegistration.cpp (hunk 17 above)
Comment (kzangeli) [reply src/lib/serviceRoutinesV2/getRegistration.cpp]: Fixed in f114c3e

File: src/lib/common/statistics.cpp (hunk 8 above)
Comment (kzangeli) [reply src/lib/common/statistics.cpp]: OK, if that part is meant to be a complete list, then one more counter, *"registrations"*, must be added.

File: src/lib/mongoBackend/CMakeLists.txt (hunk 9 above)
Comment (kzangeli) [reply src/lib/mongoBackend/CMakeLists.txt]: The S2 rule, talks about names of functions, true. However, creating a file for **one** external function, which is how it should be done, the name of the file should of course be the name of the function. So, by extrapolation, the S2 rule actually talks about files names as well.

About the 'moreover 1', you mean to not use setXXX but xxxSet?
If so, yes, I agree, of course. I used those names (that I don't like), as other files use that kind of names.

And, about 'moreover 2', of course, that file should be removed. I didn't really realize those files were they. Can't say I needed them ...

Comment (kzangeli) [reply src/lib/mongoBackend/CMakeLists.txt]: Fixed in 53980e464

File: src/lib/common/JsonHelper.cpp (hunk 6 above)
Comment (kzangeli) [reply src/lib/common/JsonHelper.cpp]: Fixed in 53980e4

I did a quick test adding a method addNumber(xxx, float).
Got a number of compilation errors, saying that addNumber was ambiguous ...
That problem for another day!

File: src/lib/common/statistics.cpp (hunk 8 above)
Comment (kzangeli) [reply src/lib/common/statistics.cpp]: Fixed in 53980e4

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 15 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Thinking a little about this ...
Normally the output params would go first, look at strcpy for example:

`char* strcpy(char *dest, const char *src)` 

Not sure I like to put all output params first, but I really like to have the 'destination' first, if a function has a destination.

Perhaps a hybrid:
```
xxx func(dest, input1, .. inputN, output2, ... outputN).
```

Secondary output params are like bieffects ...

Perhaps better to leave this be.
I would though add a rule about if a function has a destination parameter, it should be the first parameter to the function. Either that, or nothing at all. Depends on the function ...

File: test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test (hunk 19 above)
Comment (kzangeli) [reply test/functionalTest/cases/3008_get_registration/get_reg_using_v2_created_using_v1.test]: Fixed in bd588f384

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 14 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: Fixed in bd588f384

File: src/lib/common/statistics.cpp (hunk 8 above)
Comment (fgalan) [reply src/lib/common/statistics.cpp]: Well... not sure if that part was mean to be a complete list. That's why I ask for opinion :)

However, modification in 3980e4 is fine, so ok.

File: src/lib/mongoBackend/CMakeLists.txt (hunk 9 above)
Comment (fgalan) [reply src/lib/mongoBackend/CMakeLists.txt]: About "moreover 1", what I mean is that CMakeLists.txt  should be modified not only in the .cpp part, but also in the .h part.

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 15 above)
Comment (fgalan) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: If we distinguish two "categories" of output params (destination and "others") things start getting complicated and maybe the associated rule will be hard to follow. In this situation, maybe not having a rule at all... not sure.

I like the idea of having OrionError (or std::string error in some cases) at the end... the problem is that we don't follow that pattern in other cases. Could you create an issue about it, please? However, hard to detect "unaligned cases"...

File: src/lib/mongoBackend/CMakeLists.txt (hunk 9 above)
Comment (kzangeli) [reply src/lib/mongoBackend/CMakeLists.txt]: Yeah, I saw. Got compilation error about it :-)

File: src/lib/mongoBackend/mongoRegistrationGet.cpp (hunk 15 above)
Comment (kzangeli) [reply src/lib/mongoBackend/mongoRegistrationGet.cpp]: NTC for this PR. Another PR will deal with the style guide

File: CHANGES_NEXT_RELEASE [hunk 21]
```diff
@@ -1,3 +1,4 @@
 - Add: CORS Preflight Requests support for /v2 and /v2/entities, -corsMaxAge switch (#501)
 - Fix: case-sensitive header duplication (e.g. "Content-Type" and "Content-type") in custom notifications (#2893)
+- Add: support for GET /v2/registrations/<registration-id>
```
[the remark this answers is not in the document]
Comment (fgalan) [reply CHANGES_NEXT_RELEASE]: Fixed in f114c3e2dcc81a9432bc7193c4d2e2e94c4f50a0

**Comment by fgalan** [comment]
LGTM
